tensorboard/components/tf_categorization_utils/categorizationUtils.ts

```diff
@@ -17,66 +17,72 @@ import {getTags} from '../tf-backend/backend.js';
 
 /**
  * Functions to extract categories of tags and/or run-tag combinations
- * from a run-to-tag mapping, yielding data in a format suitable to be
- * used as items in a `<dom-repeat>`ed in a Polymer component.
- *
- * TODO(wchargin): Scrap `tagGroupRegexes?: string[]` and replace with a
- * `filter?: string`.
+ * from a run-to-tag mapping. The resulting categories can be fed to a
+ * `tf-category-pane`, and their items can be `<dom-repeat>`ed in a
+ * Polymer component.
  */
+
 export type RunToTag = {[run: string]: string[]};
-export type TagCategory = {
-  name: string,
-  items: {
-    tag: string,
-    runs: string[],
-  }[],
-  count: number,  // === items.length; provided for Polymer binding convenience
+
+export enum CategoryType {
+  SEARCH_RESULTS,
+  PREFIX_GROUP,
 }
-export type RunTagCategory = {
-  name: string,
-  items: {
-    tag: string,
-    run: string,
-  }[],
-  count: number,  // === items.length; provided for Polymer binding convenience
+export interface PrefixGroupMetadata {
+  type: CategoryType.PREFIX_GROUP;
+}
+export interface SearchResultsMetadata {
+  type: CategoryType.SEARCH_RESULTS;
+  validRegex: boolean;
+  universalRegex: boolean;  // is the search query ".*"? ("(?:)" doesn't count)
 }
+export type CategoryMetadata = PrefixGroupMetadata | SearchResultsMetadata;
 
-// Used internally as an intermediate structure.
-export type Category = {
+export interface Category<T> {
   name: string,
-  items: string[],
+  metadata: CategoryMetadata,
+  items: T[],
 };
+export type TagCategory = Category<{tag: string, runs: string[]}>;
+export type RunTagCategory = Category<{tag: string, run: string}>;
 
-/**
- * For each source that represents a valid regex, compute a category
- * containing all elements that match that regex. Discard invalid
- * sources.
- */
-function categorizeByRegexes(xs: string[], regexSources: string[]): Category[] {
-  return regexSources.map(source => {
+export type RawCategory = Category<string>;  // Intermediate structure.
+
+export function categorizeByFilter(xs: string[], filter: string): RawCategory[] {
```

Comment:
Also, wondering why it returns a RawCategory[] when it will always be a singleton array? I think it would be cleaner to have it return a single RawCategory, and let callers be responsible for putting it into an array